be/src/exec/pipeline/aggregate_blocking_operator.cpp

```diff
@@ -0,0 +1,124 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021 StarRocks Limited.
+
+#include "aggregate_blocking_operator.h"
+
+#include "column/chunk.h"
+#include "column/column.h"
+#include "column/column_helper.h"
+#include "column/const_column.h"
+#include "config.h"
+#include "exprs/expr_context.h"
+#include "runtime/runtime_state.h"
+
+namespace starrocks::pipeline {
+
+void AggregateBlockingOperator::finish(RuntimeState* state) {
+    if (_is_pre_finished) return;
+    _is_pre_finished = true;
+
+    if (!_group_by_expr_ctxs.empty()) {
+        COUNTER_SET(_hash_table_size, (int64_t)_hash_map_variant.size());
+        // If hash map is empty, we don't need to return value
+        if (_hash_map_variant.size() == 0) {
+            _is_finished = true;
+        }
+
+        if (false) {
+        }
+#define HASH_MAP_METHOD(NAME)                                                             \
+    else if (_hash_map_variant.type == vectorized::HashMapVariant::Type::NAME) _it_hash = \
+            _hash_map_variant.NAME->hash_map.begin();
+        APPLY_FOR_VARIANT_ALL(HASH_MAP_METHOD)
+#undef HASH_MAP_METHOD
+    } else if (_group_by_expr_ctxs.empty()) {
+        // for aggregate no group by, if _num_input_rows is 0,
+        // In update phase, we directly return empty chunk.
+        // In merge phase, we will handle it.
+        if (_num_input_rows == 0 && !_needs_finalize) {
+            _is_finished = true;
+        }
+    }
+    COUNTER_SET(_input_row_count, _num_input_rows);
+}
+
+StatusOr<vectorized::ChunkPtr> AggregateBlockingOperator::pull_chunk(RuntimeState* state) {
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+    // TODO(hcf) force annotation
+    // RETURN_IF_ERROR(exec_debug_action(TExecNodePhase::GETNEXT));
+    RETURN_IF_CANCELLED(state);
+
+    if (_is_finished) {
```

Comment:
if Operator's is_finish() return true, never invoke pull_chunk.

Reply:
_is_finished here means whether hash table has remaining data, I've changed the name so that it can be more clear